src/librustc_ast_lowering/expr.rs

```diff
@@ -818,6 +833,232 @@ impl<'hir> LoweringContext<'_, 'hir> {
         })
     }
 
+    /// Destructure the LHS of complex assignments.
+    /// For instance, lower `(a, b) = t` to `{ let (lhs1, lhs2) = t; a = lhs1; b = lhs2; }`.
+    fn lower_expr_assign(
+        &mut self,
+        lhs: &Expr,
+        rhs: &Expr,
+        eq_sign_span: Span,
+        whole_span: Span,
+    ) -> hir::ExprKind<'hir> {
+        // Return early in case of an ordinary assignment.
+        let is_ordinary = match &lhs.kind {
+            ExprKind::Array(..)
+            | ExprKind::Struct(..)
+            | ExprKind::Tup(..)
+            | ExprKind::Underscore => false,
+            ExprKind::Call(callee, ..) => {
+                // Check for tuple struct constructor.
+                if let ExprKind::Path(qself, path) = &callee.kind {
+                    let qpath = self.lower_qpath(
+                        callee.id,
+                        qself,
+                        path,
+                        ParamMode::Optional,
+                        ImplTraitContext::disallowed(),
+                    );
+                    match qpath {
+                        hir::QPath::Resolved(
+                            _,
+                            hir::Path { res: Res::Def(DefKind::Ctor(..), _), .. },
```

Comment:
We need to make sure we only allow destructuring of structs and enums that have exactly one variant and are not marked `#[non_exhaustive]`.